[WIP] Lower newExp for scalar types

edi33416 opened the pull request:
Lower calls to `new` expression to druntime's `_d_newitemT` templated implementation that uses compile time information about the type instead of `TypeInfo`.

edi33416 committed:
Lower newExp for scalar types

edi33416 commented:
@andralex @RazvanN7 @WalterBright

WalterBright commented on src/dmd/expressionsem.d:
```diff
@@ -3587,36 +3587,58 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
-                (*exp.arguments)[0] = e;
+                newExpLowering = new ConstructExp(newExpLowering.loc, newExpLowering, e.addressOf());
+                newExpLowering.expressionSemantic(sc);
             }
```
```
newExpLowering = newExpLowering.expressionSemantic(sc);
```

WalterBright replied:
You're also calling expressionSemantic() twice on the same expression.

jacob-carlborg replied:
Even this works:

```d
newExpLowering = new ConstructExp(newExpLowering.loc, newExpLowering, e.addressOf()).expressionSemantic(sc);
```

But it might be a bit too long.

WalterBright commented on src/dmd/expressionsem.d:
```diff
@@ -3587,36 +3587,58 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         }
         else if (tb.isscalar())
         {
+            Identifier id;
```
Add comment showing what the lowered expression is.

WalterBright commented on src/dmd/expressionsem.d:
```diff
@@ -3587,36 +3587,58 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
+            else
+            {
+                id = Identifier.idPool("___d_newitemiT");
+            }
```
9 lines of code to one!
```
const id = Identifier.idPool(tb.isZeroInit(exp.log) ? "__d_newitemT" : __d_newitemiT");
```

edi33416 replied:
Now this is done in the `druntime` implementation of `___d_newitemT`

jacob-carlborg commented on src/dmd/expressionsem.d:
```diff
@@ -3587,36 +3587,58 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         else if (tb.isscalar())
         {
+            Identifier id;
+            if (tb.isZeroInit(exp.loc) == true)
```
Can't the implementation of `___d_newitemT` figure this out so there's only one function that's needed?

edi33416 replied:
Moved this into `___d_newitemT`'s implementation

edi33416 committed:
Address feedback

edi33416 commented:
I addressed the feedback. This is still a WIP, as there are some tests failing in `druntime`.
I'm getting the following errors
```
src/object.d(63): Error: static variable typeid(int) cannot be read at compile time                                                                                 
src/object.d(88):        called from here: ___d_newitemU()                                                                                                          
src/core/lifetime.d(1621):        called from here: ___d_newitemT()                                                                                                 
src/core/internal/traits.d(334):        called from here: (*function () 

[...]

src/core/internal/traits.d(334):        called from here: (*function () => true)()
src/core/internal/traits.d(334):        while evaluating: static assert((*function () => true)())
src/core/lifetime.d(1606): Error: template instance `core.internal.traits.assertCTFEable!(function ()
```
How was this previously `CTFEable` and what changed now?

Another issue that I'm having is that I can't call a **non-templated** function from `___d_newitemT`. The linker will fail saying it doesn't find the function. Any ideas why this might happen?

ZombineDev commented on src/dmd/id.d:
```diff
@@ -342,6 +342,7 @@ immutable Msgtable[] msgtable =
     { "__equals"},
     { "__switch"},
     { "__switch_error"},
+    { "___d_newitemT" },
```
No need for triple underscore prefix and `T` suffix:

```suggestion
    { "__newitem" },
```

rainers replied:
I also consider double underscore to be wrong usage for regular identifiers in dmd, as these are reserved by the language specification.

There are a couple of places where an identifier behaves slightly different if it starts with `__`, e.g. https://github.com/dlang/dmd/blob/master/src/dmd/traits.d#L1392.

The mago debugger can also hide variables that start with a double underscore as these are usually compiler generated temporaries.

ZombineDev replied:
I agree with @rainers - the function are `extern (D)` and templates non the less, so we don't need to resort to legacy C-style naming practices. I think for the time being, we should at least use a name consistent with `__switch_error` and `__equals`, like `__new_item`, though `newHeapObject` (or something along those lines) would be better, considering that we probably want to lower all `new` expressions to it.

JinShil replied:
I think the problem is these symbols are in object.d which is automatically imported by all .d source files.  This could cause a naming conflict in user code.  If we could fully qualify the symbol (e.g. `object.switchError` instead of just `switchError`) in the lowering, we wouldn't need the double underscore prefix.  I don't know how to do that.  @WalterBright, do you?

rainers replied:
Although the use of double underscores is probably unnecessary in `object.d`, too, I wasn't actually referring to their use in lowerings, but their use in dmd as `Id.__newitem`. That way a reserved identifier used in lowerings appears in regular code aswell.

wilzbach replied:
>  If we could fully qualify the symbol (e.g. object.switchError instead of just switchError) in the lowering,

Sure we can and have done that even in the past ;-)
It looks like this:

https://github.com/dlang/dmd/blob/fd8eb477ee0553a0e9aaaa8284c1d1d1e299e616/src/dmd/expressionsem.d#L5663-L5666

jacob-carlborg replied:
> If we could fully qualify the symbol

That should be a requirement. I've already run into issues with a similar case where the fully qualified name wasn't used.

jacob-carlborg replied:
Regarding the double underscore. Since these need to be public symbols, the double underscore could be used to indicate that these symbols should not be used directly, only by the compiler.

JinShil commented:
> `Error: static variable typeid(int) cannot be read at compile time`

I doubt the original runtime hook was ever called at compile-time.  It's currently lowered in `e2ir` which is after all semantic and CTFE opportunity (if my understanding of the compiler is correct).  By converting this runtime hook to a template, and lowering in the semantic phase of the compiler, you've now subject the runtime template to CTFE, which has its consequences.

The best course of action is to just remove `typeid` completely from the runtime template.  As a hack, you could also add a check for `sc.flags & `SCOPE.ctfe`, but that will earn low marks.

jacob-carlborg commented on src/dmd/expressionsem.d:
```diff
@@ -3587,36 +3587,51 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         }
         else if (tb.isscalar())
         {
+            // call ___d_newitemT!(exp.newtype)()
```
Should use the fully qualified name when calling the function.

dkorpel commented:
Superseded by https://github.com/dlang/dmd/pull/14664
